### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by odisseus**

### CODE REVIEW

File: src/core/data.scala [hunk 1]
```diff
@@ -928,10 +928,22 @@ case class Schema(id: SchemaId,
...
     imports.to[List].map(resolve(_, io, layout, https)).sequence
 
-  def allProjects(io: Io, layout: Layout, https: Boolean): Try[List[Project]] =
-    importedSchemas(io, layout, https)
-      .flatMap(_.map(_.allProjects(io, layout, https)).sequence.map(_.flatten))
-      .map(_ ++ projects.to[List])
+  def allProjects(io: Io, layout: Layout, https: Boolean): Try[List[Project]] = {
+    @tailrec
+    def flatten[T](aggregated: List[T], children: T => Try[List[T]], next: List[T]): Try[List[T]] = {
+      if(next.isEmpty) ~aggregated
```
Comment (mzarnowski) [inline src/core/data.scala]: 1. is the `ch.get` an optimization? Why not use a match instead?
2. `children` parameter can be renamed  to `extract` to reflect its purpose
3. `next` could also be renamed to `sources` for the same reason.
4. Do we need to append it on the end? `:+` rebuilds the collection every time. Could we append the node at the beginning with and then, if the needs be, reverse the result of `flatten` in line 945?
5. can a schema be imported only from one source? Wouldn't this result in duplication of schemas if, for example, A and B import the same schema C?

Here is a proposal:


```scala
def flatten[T](acc: List[T], extract: T => Try[List[T]], sources: List[T]): Try[List[T]] = {
  sources match {
    case Nil =>  ~acc
    case source :: tail =>
      extract(source) match {
        case Success(children) =>
          flatten(head :: acc, children, tail ++ children)
        case failure =>
          failure
   }
}
```

File: src/core/data.scala [hunk 2]
```diff
@@ -928,10 +928,22 @@ case class Schema(id: SchemaId,
...
+      else {
+        val node = next.head
+        val ch = children(node)
+        if(ch.isSuccess) flatten(aggregated :+ node, children, next.tail ++ ch.get)
+        else ch
+      }
+    }
+
+    for {
+      allSchemas <- flatten[Schema] (Nil, _.importedSchemas(io, layout, https), List(this))
```
Comment (mzarnowski) [inline src/core/data.scala]: by making the flatten:
```scala
def flatten[A](sources: List[A])(acc: List[A], extract: A => Try[List[A]]): Try[List[A]] = ???
```

we could probably use it here like this:
```scala
for {
  allSchemas <- flatten(List(this))(Nil, _.importedSchemas(io, layout, https))
}
```
which doesn't require specifying a type

File: src/io/path.scala [hunk 3]
```diff
@@ -71,15 +71,7 @@ case class Path(value: String) {
...
-  def fileCount(pred: String => Boolean): Int =
-    Option(javaFile.listFiles).map { files =>
-      val current = files.count { f => pred(f.getName) }
-      val descendants = files.filter(_.isDirectory).map(Path(_).fileCount(pred))sum
-
-      current + descendants
-    }.getOrElse(0)
-
-  def empty: Boolean = 0 == fileCount(_ => true)
+  def empty: Boolean = Files.walk(javaPath).allMatch(p => Files.isDirectory(p))
```
Comment (mzarnowski) [inline src/io/path.scala]: `Files.walk` (and `list`) returns a stream which must be closed after consuming it.
```scala
val stream = Files.walk(javaPath)
try stream.allMatch(p => Files.isDirectory(p))
finally stream.close()
```

**Review by mzarnowski — CHANGES_REQUESTED** [review]

File: src/io/path.scala (hunk 3 above)
Comment (odisseus) [reply src/io/path.scala]: I'm not sure that's [necessary](https://stackoverflow.com/a/38698245), but let's be on the safe side.

Comment (odisseus) [reply src/io/path.scala]: Ok

File: src/core/data.scala (hunk 2 above)
Comment (odisseus) [reply src/core/data.scala]: Ok

File: src/core/data.scala (hunk 1 above)
Comment (odisseus) [reply src/core/data.scala]: Refactored

**Review by propensive — APPROVED** [review]

**Pull request merged**